### ISSUE TITLE
blast_orthofinder error

### ISSUE THREAD

**Issue by ohickl** [metagenlab/zDB#146]
Hi,

when running zDB on 105 genomes like this:
```
zdb run --input="$(basename ${INPUT_TABLE})" \
        --name="<name>" \
        --out="${OUTPUT_DIR}" \
        --ref_dir="${ZDB_REF_DIR}" \
        --cpu="${THREADS}" \
        --mem=200GB \
        --ko --cog --pfam --vfdb --swissprot --amr \
        --singularity_dir="/dev/shm" \
        --resume &> "${LOG_DIR}/zdb_run.log"
```
, I end up getting the following error:
```
...
ERROR ~ Error executing process > 'blast_orthofinder (263)'


Caused by:

  Process `blast_orthofinder (263)` terminated with an error exit status (2)



Command executed:


  blastp -outfmt 6 -evalue 0.001 -query /.../work/59/7ac0b48e8598b3b7868cabf102bc27/OrthoFinder/Results_output/WorkingDirectory/Species51.fa -db /.../work/59/7ac0b48e8598b3b7868cabf102bc27/OrthoFinder/Results_output/WorkingDirectory/BlastDBSpecies10 -num_threads 1 > Results_output/WorkingDirectory/Blast51_10.txt


Command exit status:

  2


Command output:

  (empty)


Command error:

  INFO:    Converting SIF file to temporary sandbox...

  WARNING: Skipping mount /.../singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container

  BLAST Database error: CSeqDBAtlas::MapMmap: While mapping file [/.../work/59/7ac0b48e8598b3b7868cabf102bc27/OrthoFinder/Results_output/WorkingDirectory/BlastDBSpecies10.pin] with 0 bytes allocated, caught exception:

  NCBI C++ Exception:

      T0 "/opt/conda/conda-bld/blast_1537407096784/work/c++/src/corelib/ncbifile.cpp", line 5503: Error: ncbi::CMemoryFileMap::CMemoryFileMap() - To be memory mapped the file must exist: /.../work/59/7ac0b48e8598b3b7868cabf102bc27/OrthoFinder/Results_output/WorkingDirectory/BlastDBSpecies10.pin

 

  INFO:    Cleaning up image...


Work dir:

  /.../work/d9/2e161667c4c3a453af57bc5315350c


Container:

  /.../singularity/quay.io-biocontainers-orthofinder-2.5.2--0.img


Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line 
```
I am unsure how do debug since the pin file is there and not empty after manually checking.
Also, even when using `--resume` it seems to rerun most of the steps, especially all the ~11k `blast_orthofinder` steps. Is this expected behavior?

Let me know if you need any more info.

Best

Oskar

**Comment by njohner** [comment]
The file is there but it does not get mounted into the singularity container. Did you by any chance set nextflow's `workDir` yourself to `/.../work`? Apptainer needs absolute paths to mount the files if I remember well, so here `/.../work/59/7ac0b48e8598b3b7868cabf102bc27/OrthoFinder/Results_output/WorkingDirectory/BlastDBSpecies10.pin` was not mounted.

**Comment by njohner** [comment]
Ah ok, I was confused. In that case I do not have any a priori idea why it's failing. I'll need to see whether I can reproduce the issue. Can you attache the `.nextflow.log` file (it might be one of the older ones `.nextflow.log.1` or so)? 

As to your question about the `--resume` option, this is passed directly to nextflow (as `-resume`) and that usually works well. I've just seen that for `blast_orthifinder` we use the `each` qualifier in the inputs of the process and I do not know whether that might break the resume. I'll need to test as well.

**Comment by ohickl** [comment]
Ok. I had to kill one `execute_amrscan`  (`execute_amrscan (24)`) process, since it was stuck for hours, with the run being basically done. I attached `.nextflow.log`, the stdout and the specific error log created. Not sure if the java error is a result of me killing the process.
I restarted again, but unfortunately, it seems to basically rerun everything again, as before. I attached the current log state.

[hs_err_pid3314236.zip](https://github.com/user-attachments/files/18110358/hs_err_pid3314236.zip)
[zdb_run.zip](https://github.com/user-attachments/files/18110360/zdb_run.zip)
[nextflow.log.zip](https://github.com/user-attachments/files/18110362/nextflow.log.zip)
[zdb_run_resume.zip](https://github.com/user-attachments/files/18111251/zdb_run_resume.zip)
[nextflow.log1.zip](https://github.com/user-attachments/files/18111266/nextflow.log1.zip)

**Comment by ohickl** [comment]
I got it to run by adding:
```
singularity.runOptions = "-B $baseDir:$baseDir"
singularity.envWhitelist = "PYTHONPATH"
```
to `nextflow.config`. I am running it all in RAM disk (work dir + nextflow scratch) to avoid IO/network problems. I was consistently getting this error before adding the singularity params:
```
<rest of log>
Plus 2 more processes waiting for tasks…

Annotation pipeline completed

You may now start the web server with the zdb webapp command

WARN: Got an interrupted exception while taking agent result | java.lang.InterruptedException

ERROR ~ Error executing process > 'load_base_db (1)'


Caused by:

  Process `load_base_db (1)` terminated with an error exit status (1)



Command executed:


  #!/usr/bin/env python

 

  import setup_chlamdb

 

  kwargs = "<long_list_was_here"

  gbk_list  ="<long_list_was_here"

  alignments_lst = "<long_list_was_here"


  print("Loading gbks", flush=True)

  setup_chlamdb.load_gbk(gbk_list, kwargs, "gigantic_linnaeus")

 

  print("Loading groups", flush=True)

  setup_chlamdb.load_groups("zdb_input.csv", kwargs, "gigantic_linnaeus")

 

  # kept for now, need to check whether this is really necessary to keep the hash

  print("Loading seq hashes", flush=True)

  setup_chlamdb.load_seq_hashes(kwargs, "nr_mapping.tab", "gigantic_linnaeus")

 

  print("Loading orthofinder results", flush=True)

  setup_chlamdb.load_orthofinder_results("Orthogroups.txt", kwargs, "gigantic_linnaeus")

 

  print("Loading alignments", flush=True)

  setup_chlamdb.load_alignments_results(kwargs, alignments_lst, "gigantic_linnaeus")

 

  print("Loading checkm results", flush=True)

  setup_chlamdb.load_genomes_info(kwargs, gbk_list, "checkm_results.tab", "gigantic_linnaeus")

 

  print("Loading phylogenies")

  gene_list = "<long_list_was_here"  


  setup_chlamdb.load_reference_phylogeny(kwargs, "core_genome_phylogeny.nwk", "gigantic_linnaeus")

  setup_chlamdb.load_gene_phylogenies(kwargs, "orthogroups_summary_info.tsv", gene_list, "gigantic_linnaeus")


Command exit status:

  1


Command output:

  (empty)


Command error:

  INFO:    Converting SIF file to temporary sandbox...

  Traceback (most recent call last):

    File ".command.sh", line 3, in <module>

      import setup_chlamdb

    File "/.../micromamba/envs/zdb/share/zdb-1.3.4/bin/setup_chlamdb.py", line 9, in <module>

      from lib import KO_module, search_bar

  ModuleNotFoundError: No module named 'lib'

  INFO:    Cleaning up image...


Work dir:

  /dev/shm/ohickl_zdb_work/work/3b/ee0756b9753a55a7e65afa9606cc80


Container:

  /dev/shm/registry.hub.docker.com-metagenlab-annotation-pipeline-1.4.1.img


Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`


 -- Check '.nextflow.log' file for details


Could not finish the analysis

Try to rerun the same command with the --resume flag set to rerun the analysis 
```
The problem i have now is that there are no annotations (e.g. `COG entries`) available in the UI on the webserver.
I ran it like this, skipping amrfinder, since there was always one job stuck indefinitely:
```
zdb run --input="$(basename ${INPUT_TABLE})" \
        --name="A_muciniphila" \
        --out="${OUTPUT_DIR}" \
        --ref_dir="${ZDB_REF_DIR}" \
        --cpu="${THREADS}" \
        --mem=200GB \
        --singularity_dir="/dev/shm" \
        --ko --cog --pfam --vfdb --swissprot \
        --resume &> "${LOG_DIR}/zdb_run.log"
```
Here the log of that run:
[zdb_run_v2.zip](https://github.com/user-attachments/files/18165964/zdb_run_v2.zip)

**Comment by njohner** [comment]
It's hard to know what is going from from the logs. But here a few points/suggestions:

- I have no experience with RAM disks, but if I understand well, data is copied back and forth between the RAM disk and your HD. Nextflow's resume uses hashes composed from various information, plus the files present in the `work` directory to decide whether a job has to be repeated or not. That hash notably contains the modification date of the files, so moving/copying a file will break the resume. That might be your issue there.
- I have just run zDB from scratch on an example dataset and did not encounter any issues, including running `AMRFinderPlus`. Did you check the log file of `AMRFinderPlus` process which was hanging? Each nextflow process has a corresponding log file in the `work` directory. Did you maybe run out of memory or something alike?
- As far as I know there should be no need to add any singularity options. 
- When you start the webapp, make sure to that you are opening the correct run. So in your case `zdb webapp --name=compassionate_mendel --dir="${OUTPUT_DIR}"`. 
- If you point the webapp to the correct run and it does not display the expected data, that would be very strange. You could have a look at what is in the database, or send it to me (it's in `$OUTPUTDIR/zdb/results/db`)

I could also advise to first make sure everything works using a small example dataset, and maybe simplifying things by not running on a RAM disk. As a test dataset you can for example use the one from the zDB tests (https://github.com/metagenlab/zDB/tree/master/testing/pipelines/assets), you will just need to modify the paths to the files in the `test_input.csv` file.

I hope that helps...

**Comment by ohickl** [comment]
Ok, thanks!
- The RAM disk runs cannot be resumed of course. Those are always fresh runs. I am doing this do speed things up and also to avoid crashes because my file write quota is consumed. With the 1.5K genomes, it writes hundreds of thousands of files.
- On a small subset (16 genomes) it does work fine, and also all annotations are displayed correctly in the webserver.
- I checked both the small and the large runs outputs in the webapp, the large one does correctly show e.g. the phylogeny etc, just all the functional annotations are missing. they seem to be in the db though:
```
> sqlite3 zdb_output_v2/zdb/results/db/compassionate_mendel
SQLite version 3.46.0 2024-05-23 13:25:27
Enter ".help" for usage hints.
sqlite> .tables
biodatabase                 location                  
biodb_config                location_qualifier_value  
bioentry                    og_hits                   
bioentry_dbxref             ontology                  
bioentry_path               orthology_identity        
bioentry_qualifier_value    reference                 
bioentry_reference          reference_phylogeny       
bioentry_relationship       seqfeature                
biosequence                 seqfeature_dbxref         
cog_functions               seqfeature_path           
cog_names                   seqfeature_qualifier_value
comment                     seqfeature_relationship   
dbxref                      sequence_hash_dictionnary 
dbxref_qualifier_value      taxon                     
filenames                   taxon_in_group            
gene_phylogeny              taxon_name                
genome_summary              term                      
groups                      term_dbxref               
ko_class                    term_path                 
ko_def                      term_relationship         
ko_module_def               term_relationship_term    
ko_pathway_def              term_synonym              
ko_to_module                versions                  
ko_to_pathway             
sqlite> PRAGMA integrity_check;
ok
```
here also some lines from the contained tables:
[db_inspection_results.zip](https://github.com/user-attachments/files/18179681/db_inspection_results.zip)
could it be a memory issue? i am also running through WSL on windows. but then again, with the same setup, the small output seems just fine.
what is the largest dataset you tried? i was also planning on running on an order of magnitude larger number of genomes later, if the issues can be resolved.

**Comment by njohner** [comment]
Ok I did not realize your dataset was that large. Are the genomes particularly small? Wasn't the orthofinder excessively slow? This is currently all against all, so scales as n**2, and is the clear bottleneck in the annotation pipeline. We were planing on replacing it with pirates, but haven't done that yet. zDB is typically used for datasets for up to 150 genomes.

As for the webapp, what to you mean by "functional annotations are missing"?  You have the menu points on the left hand side right? And if you navigate to `index_comp/ko` you see the available comparisons? But nothing is displayed when you go to `entry_list_ko`? zDB does the table rendering front end side, which does not work well for large tables (backend returns the whole table and the frontend sorts it and displays only the first page). So while the backend sends the data pretty quickly, then the frontend can take ages to display the data. You can verify this in the network tab of the web inspection tool (right-click in the browser and choose "inspect", which opens the inspection tools, and there choose the "network tab". reload the page and you'll see the requests. I expect the request to `entry_list_ko` to return normally, but then the data is not displayed). 
If instead in the webapp you go for example to `/extract_ko/`, i.e. the extraction form, and perform a complex search (present in a few strains and absent from a few others), does that get displayed?

So basically I think this is a performance issue. We are planing on working on the support of larger datasets, replacing orthofinder and improving table rendering by letting the backend only return the data that will be displayed, using ajax requests. So that should all happen at some point in the coming year, but it's not quite top priority just now...
